### PR TITLE
Add CFM: convex factorization machine

### PR DESCRIPTION
**Pull request by jilljenn**
De façon assez surprenante, ce modèle est plutôt mauvais, même la version ALS sur un toy dataset. Il y a peut-être une erreur dans le package.

### CODE REVIEW

**Comment by jilljenn** [comment]
On va dire que la contribution majeure, c'est le petit code pour préparer un dataset au format FM. Peut-être qu'on devrait avoir nos méthodes de preprocessing (genre retirer la moyenne).

**Comment by RaitoBezarius** [comment]
I wonder what is the best way to handle the preprocessing to feed factorization machines, one of the strongest selling points of those is the ability to carry metadata and extra features (other than the usual CF model with works / users).

At the same time, have you tried embedding other features in the FM matrix?

**Comment by jilljenn** [comment]
No, I first want to verify that everything is okay, e.g., the error should be comparable with the best existing implementation in Mangaki. Then, we can try to embed side information such as tags, and see what happens.

**Comment by RaitoBezarius** [comment]
Alright, @jilljenn

**Comment by jilljenn** [comment]
Y a un problème qui m'embête :

    You are using pip version 8.1.1, however version 9.0.1 is available.
    You should consider upgrading via the 'pip install --upgrade pip' command.

Est-il possible de mettre à jour le pip de CircleCI ? Comme ça on n'aura peut-être pas ce problème de dépendance qui fait que fastFM attend Cython.

**Comment by RaitoBezarius** [comment]
@jilljenn Yes, we can. Let us just add `pip install --upgrade pip` at the start of our CircleCI config file, but why do you need the latest version? What is the minimal version of `pip` you require?

File: mangaki/mangaki/algo/cfm.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    @property
+    def is_serializable(self):
+        return False  # Not yet
+
+    def prepare_fm(self, X):
+        nb_samples = len(X)
+        user_ids = X[:, 0]
+        work_ids = X[:, 1]
+        rows = list(range(nb_samples))
+        rows += rows
```
[the remark this answers is not in the document]
Comment (jilljenn) [reply mangaki/mangaki/algo/cfm.py]: Done. It is because we record two 1s per row.

**Comment by jilljenn** [comment]
Okay here is the thing.

fastFM waits for Cython.

On my Python 3.6, if I put both files (or only fastFM), it works well.
But for CircleCI (Python 3.5.2), even with both Cython and fastFM in `dev.txt`, I get an `ImportError`. Is the wheel broken?

**Comment by RaitoBezarius** [comment]
@jilljenn We can open an issue in the fastFM repository to query about the status of their packages. Then, pinpoint where we should attack the issue.

**Comment by jilljenn** [comment]
I think I got the answer there: https://github.com/ibayer/fastFM/pull/115
I will try to add the master version 😄

**Comment by jilljenn** [comment]
I already have the latest version. I [dropped an issue](https://github.com/ibayer/fastFM/issues/124).

**Comment by jilljenn** [comment]
C'est bon @RaitoBezarius, tu peux jeter un œil :)

File: mangaki/mangaki/algo/cfm.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    @property
+    def is_serializable(self):
+        return False  # Not yet
+
+    def prepare_fm(self, X):
+        nb_samples = len(X)
+        user_ids = X[:, 0]
+        work_ids = X[:, 1]
+        rows = list(range(nb_samples))
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/cfm.py]: Nit: `list(range(nb_samples)) * 2` would work as well, I believe.

File: mangaki/mangaki/algo/cfm.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                          shape=(nb_samples, self.nb_users + self.nb_works)
+                          ).tocsc()
+        return X_fm
+
+    def fit(self, X, y):
+        X_fm = self.prepare_fm(X)
+
+        self.chrono.save('prepare data in sparse FM format')
+
+        self.fm = als.FMRegression(n_iter=10, rank=self.rank)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/algo/cfm.py]: Please use `self` parameters rather than arbitrary ones.

File: circle.yml [hunk 4]
```diff
@@ -6,7 +6,10 @@ machine:
 dependencies:
   pre:
     - sudo apt-get update; sudo apt-get install -y software-properties-common; sudo apt-add-repository ppa:ansible/ansible -y; sudo apt-get update; sudo apt-get install -y ansible
+    - sudo apt-get install python-dev libopenblas-dev
+    - pip install --upgrade pip
     - pip install -r requirements/dev.txt
+    - pip install fastFM
```
Comment (RaitoBezarius) [inline circle.yml]: Please put it in `requirements` and let `pip` handle it.

File: mangaki/mangaki/algo/cfm.py (hunk 3 above)
Comment (jilljenn) [reply mangaki/mangaki/algo/cfm.py]: Ça ça veut dire que tu veux que je mette un `self.fm = None` dans `__init__` ?

Comment (RaitoBezarius) [reply mangaki/mangaki/algo/cfm.py]: No, you have a number of iterations parameter in `self`, use it for `n_iter`

**Comment by jilljenn** [comment]
Plz approve @RaitoBezarius!

**Pull request merged**